code/libraries/tld-parser/src/main/java/nu/marginalia/util/TldParser.java

```diff
@@ -0,0 +1,83 @@
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Parses the public suffix list into a HashSet.
+ * Provides utility functions for stripping TLDs from the domain.
+ */
+public class TldParser {
+
+    /** Parsed list of TLDs from the publicly available list */
+    public Set<String> tlds;
+
+    /**
+     * Construct TldParser by populating the hash set of tlds
+     * 
+     * @param tldPath path to location of public_suffix_list.dat
+     */
+    public TldParser(String tldPath) {
+        try {
+            __parsePublicSuffixFile(tldPath);
+        } catch (IOException e) {
+            // implement - what exception do we want to throw here?
+        }
+    }
+
+    /**
+     * Return the domain as a string without the tld, ie google.com -> google
+     * 
+     * @param domain the domain to strip the TLD from
+     * @return domain without TLD as string
+     */
+    public String stripTld(String domain) {
+        boolean match = true;
+
+        // probably a pretty naive approach, but seems simple enough -
+        // take the domain and look up each ending until one matches
+        // ex: abc.bb.com , first look up bb.com, if that does not exist, look up com
+        while (match) {
+            String lookupDomain = domain.substring(domain.indexOf(".") + 1);
+            if (tlds.contains(lookupDomain)) {
+                return domain.replace("." + lookupDomain, "");
+            }
+            if (!lookupDomain.contains(".")) {
+                match = false;
+            }
+        }
+        return domain;
+    }
+
+    /**
+     * Given a path to a valid TLD file following the format outlined in
+     * https://publicsuffix.org/list/public_suffix_list.dat, return a
+     * Set representation of all TLDs.
+     * 
+     * @throws FileNotFoundException
+     */
+    private void __parsePublicSuffixFile(String tldPath) throws IOException {
```

Comment:
If we make this method static, make it return `Set<String>`, the constructor can do this: 

`this.tlds = __parsePublicSuffixFile(...)`

I think it makes it clearer where this set is changed. Will also make testing this method easier.

Reply:
Agreed, went ahead and did this... class probably shouldnt be static (since we need to populate with the suffix list). could do DI

Reply:
There's a bunch of static singleton classes already for similar purposes.  I think in general moving these over to dependency injection (the codebase in general uses Guice) is the best choice, but since this is already such an established pattern, I think being consistent is more important.  

That said, I think down the line it makes sense to migrate these singletons over to use Guice-based injection is probably a good idea. 

For now though there's a class WmsaHome that as a bunch of static methods for looking up specific files in the install directory.  Could probably just add another one for the suffix file. Can look at how e.g. WmsaHome.getAdsDefinition() is used.   In general the tests have access to a copy of all these files.  It's created by run/setup.sh. (we should probably add a line to that file for downloading the suffix list).

Reply:
Got it. I think what i'll go ahead and do is make this class singleton for now, and it can be  changed should a larger refactor occur. As for other changes I will go ahead and do:

- [ ] Look into setup.sh to get the file and save to data folder
- [ ] Look into the WmsaHome class that has the static methods to get file locations
- [ ] Look into places where we can use this parser class (perhaps for now comments suggesting it.. will need to see the scope of the change needed)
- [ ] junit tests

Reply:
One remaining thing is to add a build.gradle file to the new library, and then reference it in the root settings.gradle.  Otherwise the build script won't build this module, and it won't let you find e.g. WmsaHome.

I think you can almost just copy `code/libraries/geo-ip/build.gradle` to `code/libraries/tld-parser/build.gradle` and trim it a bit.  It has a dependency on opencsv which you can remove, but also a dependency on `implementation project(':code:common:config')`, that should stay, it will let this module access WmsaHome.  

Later when you're looking at integrating the code, the modules that need to access this library will need to depend on `implementation project(':code:libraries:tld-parser')` in order to access this code.  

(This stuff is can be super-unintuitive and gradle in general is really bad at telling you what needs to be done, very easy to spend hours trying to debug why module A can't find module B if you don't have a good recipe for this stuff)